File: src/main/c/jromp/examples/pi_calculation_with_reduction.c
#include <omp.h>
#include <stdio.h>
#include <math.h>

double calc(double a);

int main(int argc, char *argv[]) {
    const int n = 100000;
    const double pi25dt = 3.141592653589793238462643;
    const double h = 1.0 / n;
    double result = 0.0;
    double x;

    const double initial_time = omp_get_wtime();

    #pragma omp parallel for private(x) reduction(+:result)
    for (int i = 1; i <= n; i++) {
        x = h * (i - 0.5);
        result += calc(x);
    } // Implicit barrier

    result = h * result;
    const double final_time = omp_get_wtime();

    printf("Time: %fms\n", (final_time - initial_time) * 1000);
    printf("PI is approximately: %.16f. Error: %.16f\n", result, fabs(result - pi25dt));

    return 0;
}

double calc(const double a) {
    return 4.0 / (1.0 + a * a);
}

File: src/main/java/jromp/examples/PiCalculationWithReduction.java
package jromp.examples;

import jromp.parallel.Parallel;
import jromp.parallel.operation.Operations;
import jromp.parallel.var.ReductionVariable;
import jromp.parallel.var.Variables;
import jromp.parallel.var.reduction.Sum;

public class PiCalculationWithReduction {
    public static void main(String[] args) {
        int n = 100000;
        double h = 1.0 / n;
        ReductionVariable<Double> result = new ReductionVariable<>(new Sum<>(), 0d);
        Variables vars = Variables.create().add("sum", result);

        long initialTime = System.nanoTime();

        Parallel.defaultConfig()
                .withVariables(vars)
                .parallelFor(1, n + 1, false, (id, start, end, variables) -> {
                    double x;
                    double sum = 0.0;

                    for (int i = start; i < end; i++) {
                        x = h * (i - 0.5);
                        sum += calc(x);
                    }

                    variables.<Double>get("sum").update(Operations.add(sum).get());
                })
                .join();

        double finalResult = h * result.value();
        long finalTime = System.nanoTime();

        System.out.printf("Time: %fms%n", (finalTime - initialTime) / 1e6);
        System.out.printf("PI is approximately: %.16f. Error: %.16f%n", finalResult,
                          Math.abs(finalResult - Math.PI));
    }

    private static double calc(double a) {
        return 4.0 / (1.0 + a * a);
    }
}
